atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java

```diff
@@ -258,6 +261,60 @@ public void testKeyValueEmptyRange() {
         assertTrue(list.isEmpty());
     }
 
+    @Test
+    public void testRowsColumnRange_allResultsPostFiltered() {
+        putDirect("row1", "col1", "v1", 5);
+        putDirect("row1", "col2", "v2", 5);
+
+        byte[] rowBytes = PtBytes.toBytes("row1");
+        RowColumnRangeIterator iterator = keyValueService.getRowsColumnRange(
+                TEST_TABLE,
+                ImmutableList.of(rowBytes),
+                new ColumnRangeSelection(PtBytes.EMPTY_BYTE_ARRAY, PtBytes.EMPTY_BYTE_ARRAY),
+                1,
+                1);
+        assertThat(iterator.hasNext()).isFalse();
+    }
+
+    @Test
+    public void testRowsColumnRange_postFilteredFirstPage() {
+        putDirect("row1", "col1", "v1", 5);
+        putDirect("row1", "col1", "v1a", 6);
+        putDirect("row1", "col2", "v2", 5);
+        putDirect("row1", "col3", "v3", 0);
+
+        byte[] rowBytes = PtBytes.toBytes("row1");
+        RowColumnRangeIterator iterator = keyValueService.getRowsColumnRange(
+                TEST_TABLE,
+                ImmutableList.of(rowBytes),
+                new ColumnRangeSelection(PtBytes.EMPTY_BYTE_ARRAY, PtBytes.EMPTY_BYTE_ARRAY),
+                1,
+                1);
+        assertThat(PtBytes.toBytes("v3")).isEqualTo(iterator.next().getValue().getContents());
+        assertThat(iterator.hasNext()).isFalse();
+    }
+
+    @Test
+    public void testRowsColumnRange_forwardProgressAfterValidResult() {
+        putDirect("row1", "col1", "v1a", 1);
+        putDirect("row1", "col1", "v1b", 2);
+        putDirect("row1", "col1", "v1c", 3);
+        putDirect("row1", "col1", "v1d", 4);
+        putDirect("row1", "col1", "v1e", 5);
+        putDirect("row1", "col2", "v2", 5);
+
+        byte[] rowBytes = PtBytes.toBytes("row1");
+        RowColumnRangeIterator iterator = keyValueService.getRowsColumnRange(
```

Comment:
I expected this test to fail, since we page through 1 result at a time, and the startCol will never move beyond `col1`, since each successive query should restart its search from `startTs = 6`.

However, this hits the pre-existing code path, which moves you onto the next lexicographic column name since we already found a result for startCol.

A result at `ts = 7` for `row1` would still give us forward progress, since the next range scan would start from the concrete row `row1` and `startTs = 6`, which would find `(col1, v1e)` and move us on to the next potential column.

Reply:
It's probably worth adding the test for having some results after the start, to confirm we pick the right one. I didn't see a test for this case (at least, didn't see one near the tests you've added)

```
    @Test
    public void testRowsColumnRange_abortsCorrectlyHalfway() {
        putDirect("row1", "col1", "v1a", 3);
        putDirect("row1", "col1", "v1b", 4);
        putDirect("row1", "col1", "v1c", 5);
        putDirect("row1", "col1", "v1d", 7);
        putDirect("row1", "col1", "v1e", 8);
        putDirect("row1", "col2", "v2", 5);

        byte[] rowBytes = PtBytes.toBytes("row1");
        RowColumnRangeIterator iterator = keyValueService.getRowsColumnRange(
                TEST_TABLE,
                ImmutableList.of(rowBytes),
                new ColumnRangeSelection(PtBytes.EMPTY_BYTE_ARRAY, PtBytes.EMPTY_BYTE_ARRAY),
                1,
                6);
        assertThat(PtBytes.toBytes("v1c")).isEqualTo(iterator.next().getValue().getContents());
        assertThat(PtBytes.toBytes("v2")).isEqualTo(iterator.next().getValue().getContents());
        assertThat(iterator.hasNext()).isFalse();
    }
```